### DISCUSSION TITLE
Import Error for .pye files

### DISCUSSION THREAD

**Discussion by SinolationGroup** [apache/airflow#17905] [Q&A] [answered]
I have some python files encrypted by sourcedefender -- https://pypi.org/project/sourcedefender/. These .pye files are imported in another executable file, and they are imported correctly and works fine when I run it in python directly. But when I put them into airflow as dag file, it says:

Broken DAG: [/application/airflow/dags_tmp/m2_main.py] Traceback (most recent call last):
  File "/application/airflow/dags_tmp/m2_main.py", line 30, in <module>
    from m2_func import *
  File "src/pyx/loader.pyx", line 451, in loader.MetaLoader.exec_module
ImportError

Any idea about this?

**Comment by potiuk** [comment, marked as answer]
Airflow does not (and will likely never) support encryption of the DAG files. 

Displaying code of DAGs is an essential feature of Airflow and in order for Airflow to be able to show it in the UI, it would need to be able to decrypt the files (so basically it will have to know the private key which kind of defeats the purpose of using it in the first place). 

Airflow Parser imports the files in a separate sandbox, and likely this sandbox is not properly handled by sourcedefender's hooks. Source-defender has a number of limitations when it comes to importing as I read about.

Just a side comment. By its nature, sourcedefender does not really protect you anyway. It at most obfuscates the code. I know it is used by a lot of companies, but if you know how encryption works, It just makes it harder to decrypt your code, but if you are determined enough, you could extract the private key of sourcedefender anyway and decrypt the files. But this is a different story.

**Reply by SinolationGroup** [reply]
OK, I get it. Thanks!
